### PR TITLE
Company merge event

### PR DESCRIPTION
**Pull request by ReplicaObscura**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [x] 
| Related user documentation PR URL      |  
| Related developer documentation PR URL |  
| Issue(s) addressed                     |  

#### Description:

This is a developer-only feature which adds an event (CompanyMergeEvent) which is fired both before and after companies are merged.

It includes a very simple test class for the new class which follows how similar event tests work.

Note: This PR exists because a different PR (https://github.com/mautic/mautic/pull/11229) needs this functionality. That PR is specifically about Pipedrive, however, so this PR adds the event generically. The same 2 commits from this PR are also in that PR since otherwise it would be broken, so this PR should probably be merged first.

#### Steps to test this PR:

Note: Since this is a developer-only PR, testing this feature without any additional features simply means ensuring Mautic and its existing tests don't break.

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Install/start Mautic
3. Create two companies to be merged
4. Merge the companies from within Mautic and ensure there are no errors related to the event

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11228"></a>

### CODE REVIEW

**Comment by RCheesley** [comment]
Thanks for the PR @bmcclure and welcome to the Mautic Community - I am sure that this will be useful! cc @kuzmany @escopecz for code review to make sure all is good with the PR - please do set it ready to review when it's good to be tested as currently draft status :)

File: app/bundles/LeadBundle/Event/CompanyMergeEvent.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Mautic\LeadBundle\Event;
+
+use Mautic\LeadBundle\Entity\Company;
+use Symfony\Component\EventDispatcher\Event;
+
+/**
+ * Class CompanyMergeEvent.
+ */
+class CompanyMergeEvent extends Event
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/CompanyMergeEvent.php:11]: I suggest to have value objects final. The comment has no value so I suggest to remove it.
```suggestion
final class CompanyMergeEvent extends Event
```

File: app/bundles/LeadBundle/Event/CompanyMergeEvent.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+{
+    private $victor;
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/CompanyMergeEvent.php:13]:
```suggestion
    private Company $victor;
```

File: app/bundles/LeadBundle/Event/CompanyMergeEvent.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    private $loser;
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/CompanyMergeEvent.php:15]:
```suggestion
    private Company $loser;
```

File: app/bundles/LeadBundle/Event/CompanyMergeEvent.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        $this->victor = $victor;
+        $this->loser  = $loser;
+    }
+
+    /**
+     * Returns the victor (loser merges into the victor).
+     *
+     * @return Company
+     */
+    public function getVictor()
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/CompanyMergeEvent.php:28]:
```suggestion
    /**
     * Returns the victor (loser merges into the victor).
     */
    public function getVictor(): Company
```

File: app/bundles/LeadBundle/Event/CompanyMergeEvent.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/CompanyMergeEvent.php:2]: New classes should declare strict types. Leave it as is for existing classes as adding it to an existing class could break unexpected code.
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Event/CompanyMergeEvent.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    public function getVictor()
+    {
+        return $this->victor;
+    }
+
+    /**
+     * Returns the loser (loser merges into the victor).
+     */
+    public function getLoser()
```
Comment (escopecz) [inline app/bundles/LeadBundle/Event/CompanyMergeEvent.php:36]:
```suggestion
    /**
     * Returns the loser (loser merges into the victor).
     */
    public function getLoser(): Company
```

File: app/bundles/LeadBundle/Tests/Event/CompanyMergeEventTest.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<?php
+
```
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/Event/CompanyMergeEventTest.php:2]:
```suggestion

declare(strict_types=1);

```

File: app/bundles/LeadBundle/Tests/Event/CompanyMergeEventTest.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+namespace Mautic\LeadBundle\Tests\Event;
+
+use Mautic\LeadBundle\Entity\Company;
+use Mautic\LeadBundle\Event\CompanyMergeEvent;
+use PHPUnit\Framework\TestCase;
+
+class CompanyMergeEventTest extends TestCase
+{
+    public function testConstructGettersSetters()
```
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/Event/CompanyMergeEventTest.php:11]:
```suggestion
    public function testConstructGettersSetters(): void
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Thanks for the PR! It looks good. As Mautic supports PHP 7.4 as minimal PHP version we can use types and our static analysis checks force us to do so. I'm suggesting mostly the same changes as PHPSTAN and CS Fixer. Take a look at the failing CI jobs. It will show you what else can be improved.

**Comment by kuzmany** [comment]
Close favor to https://github.com/mautic/mautic/pull/11748

**Pull request closed without merging**